dbms/src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -1156,22 +1156,21 @@ void ExpressionAnalyzer::normalizeTreeImpl(
         {
             if (typeid_cast<ASTAsterisk *>(asts[i].get()))
             {
-                ASTs all_columns;
+                Names all_columns_name;
 
-                if (storage)
-                {
-                    /// If we select from a table, get only not MATERIALIZED, not ALIAS columns.
-                    for (const auto & name_type : storage->getColumns().ordinary)
-                        all_columns.emplace_back(std::make_shared<ASTIdentifier>(name_type.name));
-                }
-                else
-                {
-                    for (const auto & name_type : source_columns)
-                        all_columns.emplace_back(std::make_shared<ASTIdentifier>(name_type.name));
-                }
+                auto columns_name = storage ? storage->getColumns().ordinary.getNames() : source_columns.getNames();
+                all_columns_name.insert(all_columns_name.begin(), columns_name.begin(), columns_name.end());
+
+                NameSet joined_columns;
+                collectJoinedColumns(joined_columns);
```

Comment:
It's better not to call `collectJoinedColumns` from here because it has side effects.
Actually, we need only `columns_from_joined_table`. It's created in `collectJoinedColumns` method, and this code may be extracted to `AnalyzedJoin`'s member function. Also, it's better to add bool flag in order to check that  `columns_from_joined_table` won't be calculated twice.

Reply:
I can try to finish this part of the work : )